crates/pallet-domains/src/lib.rs

```diff
@@ -2020,8 +2049,13 @@ impl<T: Config> Pallet<T> {
             .and_then(|mut runtime_object| runtime_object.raw_genesis.take_runtime_code())
     }
 
-    pub fn domain_best_number(domain_id: DomainId) -> Option<DomainBlockNumberFor<T>> {
-        Some(HeadDomainNumber::<T>::get(domain_id))
+    pub fn domain_best_number(domain_id: DomainId) -> Result<DomainBlockNumberFor<T>, BundleError> {
+        // The missed domain runtime upgrades will derive domain blocks thus should be accountted
```

Comment:
Does each upgrade derive a separate block? Can there be 2 in a block?

Reply:
Actually, why can't they be counted in `HeadDomainNumber` directly?

Reply:
>Does each upgrade derive a separate block?

Not really a separate block, a consensus block that contains a domain runtime upgrade will derive a domain block no matter if it contains a bundle or not

>Can there be 2 in a block?

It won't because the key of the storage map `ScheduledRuntimeUpgrades` is the consensus block number, if a consensus block contains 2 `upgrade_domain_runtime` extrinsic, the second one will replace the first one.

>Actually, why can't they be counted in HeadDomainNumber directly?

This is because the number of domain instances that are associated with the same runtime object is unbounded, we can't update the `HeadDomainNumber` for all of them so we use a lazy approach.